sam_learning/learners/sam_learning.py

```diff
@@ -218,6 +223,44 @@ def deduce_initial_inequality_preconditions(self) -> None:
                 if action_data.signature[lifted_param1] == action_data.signature[lifted_param2]:
                     action_data.preconditions.root.inequality_preconditions.add((lifted_param1, lifted_param2))
 
+    def hard_remove_negative_preconditions(self):
+        """Removes all negative preconditions"""
+        for action in self.partial_domain.actions.values():
+            new_preconditions = set()
+
+            for precondition in action.preconditions.root.operands:
+                if isinstance(precondition, Predicate) and not precondition.is_positive:
+                    continue
+
+                new_preconditions.add(precondition)
+
+            action.preconditions.root.operands = new_preconditions
+
+    def soft_remove_negative_preconditions(self):
+        """Removes negative preconditions when the fluent is an add effect"""
+        for action in self.partial_domain.actions.values():
+            new_preconditions = set()
+            action_add_effects = [effect.untyped_representation for effect in action.discrete_effects if effect.is_positive]
+
+            for precondition in action.preconditions.root.operands:
+                if isinstance(precondition, Predicate) and not precondition.is_positive:
+                    copy_precondition_positive = precondition.copy()
+                    copy_precondition_positive.is_positive = True
+                    if copy_precondition_positive.untyped_representation in action_add_effects:
+                        continue
+
+                new_preconditions.add(precondition)
+
+            action.preconditions.root.operands = new_preconditions
+
+    def handle_negative_preconditions_policy(self):
+        match self.negative_preconditions_policy:
```

Comment:
Notice that switch case was only added in python 3.10 (as far as I know) and our code supports python 3.9 (at lease that's how I run the code)

Reply:
I checked the readme doc first before using, and it says python 3.10+

Reply:
Really?  But my code is python 3.9.12, so it should work with my environment as well, I'll fix the readme.